tests/udf/test_valid_mask.py

```diff
@@ -0,0 +1,475 @@
+import numpy as np
+import pytest
+
+from libertem.api import Context
+from libertem.common.shape import Shape
+from libertem.udf.base import UDF, MergeAttrMapping
+from libertem.io.dataset.memory import MemoryDataSet
+from libertem.common.buffers import (
+    BufferWrapper, InvalidMaskError, get_inner_slice, get_bbox, get_bbox_slice,
+)
+from libertem.common.math import count_nonzero, prod
+
+
+class ValidNavMaskUDF(UDF):
+    def __init__(self, debug=True):
+        super().__init__(debug=debug)
+
+    def get_result_buffers(self):
+        return {
+            'buf_sig': self.buffer(kind='sig', dtype=np.float32),
+            'buf_nav': self.buffer(kind='nav', dtype=np.float32),
+            'buf_single': self.buffer(kind='single', dtype=np.float32, extra_shape=(1,)),
+        }
+
+    def get_results(self):
+        assert self.meta.get_valid_nav_mask() is not None
+        assert self.meta.get_valid_nav_mask().sum() > 0, \
+            "get_results is not called with an empty valid nav mask"
+        assert len(self.meta.get_valid_nav_mask().shape) == 1, \
+            "valid_nav_mask should be flattened"
+        if self.meta.roi is not None:
+            assert self.meta.get_valid_nav_mask().shape[0] == np.count_nonzero(self.meta.roi), \
+                "if a `roi` is given, the valid nav mask should be compressed to it by default"
+            full_mask = self.meta.get_valid_nav_mask(full_nav=True)
+            assert full_mask.shape[0] == self.meta.dataset_shape.nav.size, \
+                "when passing `full_nav=True`, the shape must match the flattened ds shape"
+        if self.params.debug:
+            print("get_results", self.meta.get_valid_nav_mask())
+        results = super().get_results()
+        # import pdb; pdb.set_trace()
+        return results
+
+    def process_frame(self, frame):
+        assert self.meta.get_valid_nav_mask() is None
+        assert self.meta.get_valid_nav_mask(full_nav=True) is None
+        self.results.buf_sig += frame
+        self.results.buf_nav[:] = frame.sum()
+        self.results.buf_single[:] = frame.sum()
+
+    def merge(self, dest, src):
+        assert self.meta.get_valid_nav_mask() is not None
+        assert not np.allclose(True, self.meta.get_valid_nav_mask()), \
+            "valid nav mask should be the already merged positions! can't be all-True"
+        if self.params.debug:
+            print("merge", self.meta.get_valid_nav_mask())
+        dest.buf_sig += src.buf_sig
+        dest.buf_single += src.buf_single
+        dest.buf_nav[:] = src.buf_nav
+
+
+def test_valid_nav_mask_available():
+    dataset = MemoryDataSet(datashape=[16, 16, 32, 32], num_partitions=4)
+    ctx = Context.make_with('inline')
+    for res in ctx.run_udf_iter(dataset=dataset, udf=ValidNavMaskUDF()):
+        # TODO: maybe compare damage we got in `get_results` with `res.damage` here?
+        pass
```

Comment:
Is there a reason to not use the default fixtures for dataset and inline context?

The pattern repeats in the following tests.

Reply:
For one, I've lost overview over the different fixtures we have available. We may need documentation for the most commonly used fixtures in our contributor docs... I used `make_with` as it's just so convenient :) I'll update the tests to use `lt_ctx` instead, which actually does useful things for testing (replaces the plotting with a dummy, and restricts the number of inline threads).

As for the data set, I think it's good to actually have the code in the test, as that means it is self-contained, and you can directly understand the shape of the test data from the test code. When using the fixture, you need to go hunting first. Let me know if you have concerns with this approach, I can easily update the tests to use the `default_raw` fixture instead.

Reply:
Good point about the data shape! I also really like `make_with()`. :-)